quil-rs/src/program/memory.rs

```diff
@@ -67,323 +103,150 @@ pub enum MemoryAccessType {
-                ..Default::default()
-            },
-            Instruction::Store(Store {
-                destination,
-                offset,
-                source,
-            }) => {
-                let mut reads = vec![&offset.name];
-                if let Some(source) = source.get_memory_reference() {
-                    reads.push(&source.name);
-                }
-                MemoryAccesses {
-                    reads: set_from_reference_vec![reads],
-                    writes: set_from_reference_vec![vec![destination]],
-                    ..Default::default()
                 }
             }
-            Instruction::Declaration(_)
-            | Instruction::Fence(_)
-            | Instruction::FrameDefinition(_)
-            | Instruction::Halt()
-            | Instruction::Wait()
-            | Instruction::Include(_)
-            | Instruction::Jump(_)
-            | Instruction::Label(_)
-            | Instruction::Nop()
-            | Instruction::Pragma(_)
-            | Instruction::Reset(_)
-            | Instruction::SwapPhases(_)
-            | Instruction::WaveformDefinition(_) => Default::default(),
-        })
-    }
-}
 
-impl ArithmeticOperand {
-    pub fn get_memory_reference(&self) -> Option<&MemoryReference> {
-        match self {
-            ArithmeticOperand::LiteralInteger(_) => None,
-            ArithmeticOperand::LiteralReal(_) => None,
-            ArithmeticOperand::MemoryReference(reference) => Some(reference),
+            // We've finished our traversal; there are no subexpressions left.
+            None
         }
     }
+
+    impl std::iter::FusedIterator for MemoryReferences<'_> {}
 }
 
 impl Expression {
-    /// Return, if any, the memory references contained within this Expression.
-    pub fn get_memory_references(&self) -> Vec<&MemoryReference> {
-        match self {
-            Expression::Address(reference) => vec![reference],
-            Expression::FunctionCall(FunctionCallExpression { expression, .. }) => {
-                expression.get_memory_references()
-            }
-            Expression::Infix(InfixExpression { left, right, .. }) => {
-                let mut result = left.get_memory_references();
-                result.extend(right.get_memory_references());
-                result
-            }
-            Expression::Number(_) => vec![],
-            Expression::PiConstant() => vec![],
-            Expression::Prefix(PrefixExpression { expression, .. }) => {
-                expression.get_memory_references()
-            }
-            Expression::Variable(_) => vec![],
-        }
+    /// Return an iterator over all the memory references contained within this expression.
+    pub fn memory_references(&self) -> expression::MemoryReferences<'_> {
+        expression::MemoryReferences { stack: vec![self] }
     }
 }
 
 impl WaveformInvocation {
     /// Return, if any, the memory references contained within this WaveformInvocation.
```

Comment:
While you're here anyway...

```suggestion
    /// Return, if any, the memory references contained within this `WaveformInvocation`.
```